### ISSUE TITLE
CFP: Measure connections/sec rate in the datapath

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#20068]
## Cilium Feature Proposal

**Is your feature request related to a problem?**

Yes, under high load the connection tracking map can become full. Typically the LRU property of the map will mitigate some degree of high load on the CT map to a point. However, if the connection rate is too high then the LRU property may bump out CT entries corresponding to live connections. When this occurs, it is possible for replies for active connections to fail the CT lookup, then be treated as "forward" direction flows outbound, which are not allowed by the policy. These replies are then dropped, which begins to impact the active connections.

**Describe the feature you'd like**

This proposal is to gain better observability insights into the connection rate going into Cilium, so that it is possible to test load rates that Cilium can handle, and potentially tweak settings such as CT map sizes based upon the active load.

**(Optional) Describe your proposed solution**

Annotate the [`ct_create4()`](https://github.com/cilium/cilium/blob/48902c1469462ded5e78ea76ae9d37b1e0c9f8a4/bpf/lib/conntrack.h#L915) / `ct_create6()` success conditions to capture metrics to the [metrics map](https://github.com/cilium/cilium/blob/48902c1469462ded5e78ea76ae9d37b1e0c9f8a4/bpf/lib/metrics.h#L24) for newly created connections. Ensure that these are exported to prometheus, eg by observing the mappings in `cilium metrics list`.

Consider counting the metrics against which endpointID the connections are created for. These are likely not reasonable to export beyond the node via prometheus, but could provide additional insights even if they are only collected and presented on the local cilium-agent API.

**Comment by zhanghe9702** [comment]
@joestringer , as my unstanding, is it using an reason code < dropMin , then define two prometheus counter for ipv4/ipv6, let prometheus library function rate()/irate() show this metircs?

**Comment by joestringer** [comment]
Hi @zhanghe9702 I'm not sure what you mean by `reason code < dropMin`, but yes the idea would be to export this metric up through the Cilium metrics API (prometheus) so that the user can use rate()/irate() against these metrics. Probably this means it makes sense for the counters to be monotonically increasing. I'm not sure if it needs to be two counters for ipv4/ipv6 or just to label the ipv4/ipv6 on the same metric - for that, I would just follow the style used by other similar metrics already in Cilium.

**Comment by MadMax240484** [comment]
Hi. Do you have any update for this case?
Will it be correct to calculate cps just to use func rate for conntrack_gc_entries?

**Comment by joestringer** [comment]
I don't think that calculating against `conntrack_gc_entries` would work very well in any existing version of Cilium (up to and including v1.17.x) as this metric is only updated when garbage collection happens. Depending on your environment this could be hours between updates.
